utility/sourcemap_accessor.go

```diff
@@ -0,0 +1,190 @@
+
+	"github.com/elastic/beats/libbeat/common"
+	"github.com/elastic/beats/libbeat/logp"
+	"github.com/elastic/beats/libbeat/outputs/elasticsearch"
+)
+
+type SmapAccessor interface {
+	Fetch(smapId SmapID) (*sourcemap.Consumer, error)
+	RemoveFromCache(smapId SmapID)
+}
+
+type Enum string
+
+const (
+	InitError   Enum = "InitError"
+	AccessError Enum = "AccessError"
+	MapError    Enum = "MapError"
+	ParseError  Enum = "ParseError"
+)
+
+type SmapError struct {
+	Msg  string
+	Kind Enum
+}
+
+func (e SmapError) Error() string {
+	return fmt.Sprintf("%s", e.Msg)
+}
+
+type SmapID struct {
+	ServiceName    string
+	ServiceVersion string
+	Path           string
+}
+
+type SmapConfig struct {
+	CacheExpiration      time.Duration //seconds
+	CacheCleanupInterval time.Duration //seconds
+	ElasticsearchConfig  *common.Config
+	Index                string
+}
+
+type SourcemapAccessor struct {
+	esClients              []elasticsearch.Client
+	smapCache              *cache.Cache
+	cacheDefaultExpiration time.Duration
+	index                  string
+}
+
+func NewSourcemapAccessor(config SmapConfig) (*SourcemapAccessor, error) {
+	logp.Debug("sourcemap", "NewSourcemapAccessor created at Time.now %v for index", time.Now().Unix(), config.Index)
+	esClients, err := elasticsearch.NewElasticsearchClients(config.ElasticsearchConfig)
+	if err != nil || esClients == nil || len(esClients) == 0 {
+		err := SmapError{Msg: "Sourcemap ESClient cannot be initialized.", Kind: InitError}
+		logp.Err(err.Error())
+		return nil, err
+	}
+
+	smapCache := cache.New(config.CacheExpiration, config.CacheCleanupInterval)
+	return &SourcemapAccessor{
+		esClients:              esClients,
+		smapCache:              smapCache,
+		cacheDefaultExpiration: config.CacheExpiration,
+		index: fmt.Sprintf("%v*", config.Index),
+	}, nil
+}
+
+func (s *SourcemapAccessor) Fetch(smapId SmapID) (*sourcemap.Consumer, error) {
+	smapConsumer := s.fetchFromCache(smapId)
+	if smapConsumer != nil {
+		return smapConsumer, nil
+	}
+	smapConsumer, err := s.fetchFromES(smapId)
+	if err != nil {
+		return nil, err
+	}
+	s.AddToCache(smapId, smapConsumer)
+	return smapConsumer, nil
+}
+
+func (s *SourcemapAccessor) AddToCache(smapId SmapID, smap *sourcemap.Consumer) {
+	s.smapCache.Set(smapId.key(), smap, s.cacheDefaultExpiration)
+	logp.Debug("sourcemap", "Added smapId %v. Cache now has %v entries.", smapId.key(), s.smapCache.ItemCount())
+}
+
+func (s *SourcemapAccessor) RemoveFromCache(smapId SmapID) {
+	s.smapCache.Delete(smapId.key())
+	logp.Debug("sourcemap", "Removed smapId %v. Cache now has %v entries.", smapId.key(), s.smapCache.ItemCount())
+}
+
+func (s *SourcemapAccessor) fetchFromCache(smapId SmapID) *sourcemap.Consumer {
+	if cached, found := s.smapCache.Get(smapId.key()); found {
+		return cached.(*sourcemap.Consumer)
+	}
+	return nil
+}
+
+func (s *SourcemapAccessor) fetchFromES(smapId SmapID) (*sourcemap.Consumer, error) {
+	body := map[string]interface{}{
```

Comment:
not a huge deal but this query could just pull the sourcemap `"_source": "sourcemap.sourcemap"`